Add Dockerfile for running raiko-host and raiko-guest

pbeza opened the pull request.

pbeza committed:
Early version of raiko Docker environment

pbeza committed:
Empty reply from server

pbeza committed:
Fix Docker issues

pbeza committed:
Cleanup

pbeza commented:
@johntaiko can you please review and check if that works for you? Just run:
```
docker-compose run --rm raiko
```
It starts `raiko-host` using port `9090` by default. You can send JSON requests to the running container with:
```
curl --location --request POST 'http://<CONTAINER_IP>:9090/' --header 'Content-Type: application/json' --data-raw '{                                         
  "jsonrpc": "2.0",                                                                                                                                                                                                                                                                                                                                                                   
  "id": 1,                                                                                                                                                                                
  "method": "proof",                                                                                                                                                                      
  "params": [                                                                                                                                                                             
    {                                                                                                                                                                                     
      "type": "Sgx",                                                                                                                                                                      
      "l2Rpc": "https://rpc.internal.taiko.xyz",                                             
      "l1Rpc": "https://l1rpc.internal.taiko.xyz",                                           
      "block": 2,                                                                                                                                                                         
      "prover": "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",                                
      "graffiti": "0000000000000000000000000000000000000000000000000000000000000000"                                                                                                      
    }                                                                                                                                                                                                                                                                                                                                                                                 
  ]                                                                                                                                                                                       
}'
```

pbeza committed:
Mount logs dir as a Docker volume

johntaiko commented on docker/Dockerfile:
```diff
@@ -0,0 +1,70 @@
+FROM ${UBUNTU_IMAGE}
+
+ARG UBUNTU_CODENAME=jammy
+WORKDIR /usr/src/raiko
```
I think you can't place raiko in `/usr` https://tldp.org/LDP/Linux-Filesystem-Hierarchy/html/usr.html
`/opt/raiko` maybe be a good place https://tldp.org/LDP/Linux-Filesystem-Hierarchy/html/opt.html

johntaiko commented on docker/Dockerfile:
```diff
@@ -0,0 +1,70 @@
+ARG UBUNTU_CODENAME=jammy
+WORKDIR /usr/src/raiko
+
+ENV IP_NUMBER=0.0.0.0
```
```suggestion
ENV BIND=0.0.0.0:9090
```

johntaiko commented on docker/Dockerfile:
```diff
@@ -0,0 +1,70 @@
+ENV IP_NUMBER=0.0.0.0
+ENV PORT_NUMBER=9090
+ENV SGX_INSTANCE_ID=123
+ENV LOG_PATH=/data/log/sgx
```
We can add a config file instead of environment for arguments

smtmfft replied:
can use .env to share the vars between compose & dockerfile. I think at least the ports need export in compose.yml

johntaiko replied:
We need to frequently modify configurations and the number of configuration items will continue to increase.
So, it's not a good idea to place our configurations in env.

johntaiko commented on docker/Dockerfile:
```diff
@@ -0,0 +1,70 @@
+    cd raiko-host/guests/sgx && \
+    gramine-sgx ./raiko-guest bootstrap && \
+    cd - && \
+    RUST_LOG=debug cargo run --bin raiko-host -- --sgx-instance-id=${SGX_INSTANCE_ID} --bind=${IP_NUMBER}:${PORT_NUMBER} --log-path=${LOG_PATH}
```
`RUST_LOG=debug` Log level needs to be configurable, and we must move the final compiled executions into a new image environment without `rust` things.

johntaiko commented on docker/Dockerfile:
```diff
@@ -0,0 +1,70 @@
+
+CMD /restart_aesm.sh && \
+    cd raiko-host/guests/sgx && \
+    gramine-sgx ./raiko-guest bootstrap && \
```
We can't run bootstrap on each run

johntaiko replied:
Maybe we can let users pass the private key themselves.

johntaiko committed:
Add Dockerfile for building Gramine

johntaiko committed:
Update Dockerfile environment variables

johntaiko committed:
Update Dockerfile to support build flags

johntaiko committed:
Update Dockerfile path in docker-compose.yml

johntaiko committed:
Update Dockerfile and manifest template

This commit updates the Dockerfile to use the latest Gramine image as the base and makes changes to the installation process. It also updates the manifest template to use the correct versions of libssl and libcrypto.

johntaiko committed:
Update Dockerfile and raiko-guest manifest template

johntaiko commented on raiko-guest/config/raiko-guest.manifest.template:
```diff
@@ -37,8 +37,8 @@ sgx.trusted_files = [
   "file:{{ arch_libdir }}/libgcc_s.so.1",
-  "file:{{ arch_libdir }}/libssl.so.3",
-  "file:{{ arch_libdir }}/libcrypto.so.3",
+  "file:{{ arch_libdir }}/libssl.so.1.1",
```
@pbeza I don't know if these versions need to be 3 exactly

johntaiko committed:
Update Dockerfile and .dockerignore

johntaiko commented on Dockerfile:
```diff
@@ -0,0 +1,42 @@
+    libclang-dev
+RUN cargo build --release ${BUILD_FLAGS}
+
+FROM gramineproject/gramine:1.6-focal as runtime
```
Reuse the gramine image if we can split builder and runtime

johntaiko committed:
Refactor Dockerfile to improve directory structure and file paths

johntaiko committed:
Update secrets directory path in manifest template

johntaiko committed:
Update Dockerfile and docker-compose.yml

This commit updates the Dockerfile and docker-compose.yml files. Changes include removing unnecessary environment variables in the Dockerfile and adding new volume mounts in the docker-compose.yml file.

johntaiko commented on Dockerfile:
```diff
@@ -0,0 +1,37 @@
+    gramine-sgx-sign --manifest raiko-guest.manifest --output raiko-guest.manifest.sgx && \
+    cd -
+
+ENTRYPOINT [ "/opt/raiko/bin/raiko-host" ]
```
Remove bootstrap, we don't want to bootstrap each start

johntaiko commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,18 @@
+    volumes:
+      - /data/log/sgx:/data/log/sgx
+      - /tmp/sgx:/tmp/sgx
+      - /secrets:/opt/raiko/secrets
```
Get bootstrap secret from user

pbeza committed:
Address code review comments

pbeza committed:
Make Dockerfile work again!

pbeza committed:
Final touches before testnet

Run with:
$ docker-compose run --rm raiko --init
$ docker-compose up raiko

pbeza commented:
@johntaiko @smtmfft I think you can merge it now. Alternatively feel free to review it once again. Just please don't introduce any untested code changes anymore. Thanks!

To init/bootstrap SGX, run:
```
$ docker-compose run --rm raiko --init
```
Once bootstrapped, you can run `raiko-host` with:
```
$ docker-compose up raiko
```
Test with:
```
curl --location --request POST 'http://<CONTAINER_IP>:8085' --header 'Content-Type: application/json' --data-raw '{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "proof",
  "params": [
    {
      "type": "Sgx",
      "l2Rpc": "https://rpc.internal.taiko.xyz",
      "l1Rpc": "https://l1rpc.internal.taiko.xyz",
      "block": 2,
      "prover": "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
      "graffiti": "0000000000000000000000000000000000000000000000000000000000000000"
    }
  ]
}'
```
Just make sure to replace `CONTAINER_IP` with the IP of your container. You can check it with `docker ps` and [this magic command](https://stackoverflow.com/a/20686101/1321680):
```
docker inspect \
  --format '{{ .NetworkSettings.IPAddress }}' container_name_or_id
```
(there must be an easier way to interact with Docker container...)
